exporter/splunkhecexporter/client.go

```diff
@@ -36,25 +37,145 @@ import (
 	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/splunk"
 )
 
+var (
+	errOverCapacity = errors.New("over capacity")
+)
+
+// Minimum number of bytes to compress. 1500 is the MTU of an ethernet frame.
+const minCompressionLen = 1500
+
 // client sends the data to the splunk backend.
 type client struct {
-	config  *Config
-	url     *url.URL
-	client  *http.Client
-	logger  *zap.Logger
-	zippers sync.Pool
-	wg      sync.WaitGroup
-	headers map[string]string
+	config         *Config
+	url            *url.URL
+	client         *http.Client
+	logger         *zap.Logger
+	wg             sync.WaitGroup
+	headers        map[string]string
+	gzipWriterPool *sync.Pool
 }
 
 // bufferState encapsulates intermediate buffer state when pushing data
 type bufferState struct {
-	buf      *bytes.Buffer
-	tmpBuf   *bytes.Buffer
-	bufFront *index
-	bufLen   int
-	resource int
-	library  int
+	compressionAvailable bool
+	compressionEnabled   bool
+	bufferMaxLen         uint
+	writer               io.Writer
+	buf                  *bytes.Buffer
+	bufFront             *index
+	resource             int
+	library              int
+	gzipWriterPool       *sync.Pool
+}
+
+func (b *bufferState) reset() {
+	b.buf.Reset()
+	b.compressionEnabled = false
+	b.writer = &cancellableBytesWriter{innerWriter: b.buf, maxCapacity: b.bufferMaxLen}
+}
+
+func (b *bufferState) Read(p []byte) (n int, err error) {
+	return b.buf.Read(p)
+}
+
+func (b *bufferState) Close() error {
+	if _, ok := b.writer.(*cancellableGzipWriter); ok {
+		return b.writer.(*cancellableGzipWriter).close()
+	}
+	return nil
+}
+
+// accept returns true if data is accepted by the buffer
+func (b *bufferState) accept(data []byte) bool {
```

Comment:
It always rejects when `b.buf.Len()` is less than minCompressionLen and the received data length is greater than `bufCap`. `cancellableBytesWriter` will return `errOverCapacity` and switch over to a zip buffer condition won't be true as `b.buf.Len()` is unchanged.

Reply:
I think I fixed this with the next commit, please take a look?

Reply:
Yes, it is working now. Thanks.